07_errors/kasunfdo/error.go

```diff
@@ -0,0 +1,25 @@
+package main
+
+import "fmt"
+
+const (
+	ErrInvalidVal = 400
+	ErrNotFound   = 404
+	ErrInternal   = 500
+	ErrOpenDB     = 501
+)
+
+var ErrStr = map[int]string{
+	ErrInvalidVal: "value cannot be negative",
+	ErrNotFound:   "puppy with id=%v not found",
+	ErrInternal:   "internal system error",
+	ErrOpenDB:     "error occurred while opening db %v",
+}
+
+func NewError(code int, args ...interface{}) *Error {
+	return &Error{Code: code, Message: fmt.Sprintf(ErrStr[code], args...)}
```

Comment:
IMO this should be something like the following (doesn't compile as Cause is only partially folded in)/.
```go
type ErrCode uint8

const (
	ErrInvalidInput ErrCode = 400
        // ...
)

func (e ErrCode) String() string {
	switch e {
	case ErrInvalidInput:
		return "invalid input"
         // ....
         }
}

NewError(code int, msg string, args ...interface{}) *Error {
     return &Error{
		Message: fmt.Sprintf(format, args...),
		Code:    code,
    }
}

type Error struct {
	Message string
	Code    ErrCode
	Cause   error // can be useful if wrapping underlying error
}

func (e *Error) Error() string {
	msg := fmt.Sprintf("%s: %s", e.Code.String(), e.Message)
	if e.Cause == nil {
		return msg
	}
	return fmt.Sprintf("%s\n\t%s", msg, e.Cause.Error())
}

```